Allow configuring nginx worker reload behaviour, to prevent multiple concurrent worker reloads which can lead to high resource usage and OOMKill

rsafonseca opened the pull request:
## What this PR does / why we need it:
This PR introduces the ability to configure the concurrency behaviour of chained configuration updates that require a backend reload.

It does so by requeuing jobs that require a backend reload, whenever there's already a reload in progress, and verifies if the reload is finished by counting the number of workers, as per the configuration.

The default behaviour remains unchanged for now, and this behaviour can be configured via setting an option in the configmap.

The main purpose is to prevent several worker reloads to happen before the previous reload has finished, in order to avoid a steep and uncontrolled growth of resource usage, which can lead to OOMKill of the ingress-controller pods and/or CPU starvation. This is especially problematic in environments that use long lived connections (since the workers stay alive until the worker-shutdown-timeout is reached) and where a few configuration reloads might be triggered within that timeout.

There's a few issues regarding this problem such as #8166 but the solutions presented introduce other problems. The sync throttling suggested in the linked issue for example prevents service upstream endpoints from being updated on time, even if they don't require a reload, since all configuration changes currently share the same queue, hence the same throttling (for example, in the configuration proposed it could take up to ~2 minutes to remove terminating endpoints from a service, leading to HTTP 502)

This method addresses the root cause of the problem: multiple worker reloads which can happen spontaneously causing a huge amount of resources to be used, leading to OOMKill.

A few other related issues: #8336 #8362 #8357

## Types of changes
- [x] New feature (non-breaking change which adds functionality)

## How Has This Been Tested?
Running in a live cluster, with and without the option enabled.

## Checklist:
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I've read the [CONTRIBUTION](https://github.com/kubernetes/ingress-nginx/blob/main/CONTRIBUTING.md) guide
- [ ] I have added unit and/or e2e tests to cover my changes.
- [x] All new and existing tests passed.

rikatz commented:
/ok-to-test

rikatz commented on internal/ingress/controller/nginx.go:
```diff
@@ -668,6 +672,11 @@ Error: %v
 func (n *NGINXController) OnUpdate(ingressCfg ingress.Configuration) error {
+	concurrentlyReloadWorkers := n.store.GetBackendConfiguration().ConcurrentlyReloadWorkers
+	if !concurrentlyReloadWorkers && n.workersReloading {
+		return errors.New("worker reload already in progress, requeuing reload")
```
should this be an error, or just a warning?

rsafonseca replied:
The idea of returning an error is to pick up on the existing behaviour on the queue, which will re-queue in case of error here https://github.com/kubernetes/ingress-nginx/blob/8227888ba0a9b09afb4514284ef89f331d4c60b9/internal/task/queue.go#L130

rikatz commented:
@rsafonseca first of all, thank you for taking care of this long standing problem, this is great!

As a suggestion, I'm not sure if using a single variable is the right approach here, you may still hit some issue on multiple reloads or something that causes a different behavior to replace the variable.

Instead, and just thinking loud (I know it may not be an easy approach) what if when this feature is enabled, define a buffered channel for the reload that controls, as a semaphore, the reloads?

I'm still not sure if it does make sense, but some channel that accepts just one item, and a reload would add an item on the queue (and block), and the checker process of workers that removes this item from the channel, unblocking any other attempts to reload.

You can then control if a reload was already requested checking the length of this channel. 

@tao12345666333 you may have some better view then me here, as I honestly sucks on concurrency problems :)

rikatz commented:
/triage accepted
/priority important-soon

rikatz commented:
Also, another suggestion: As this is feature flagged (thanks!!!) let's try using positive behavior flag. Maybe instead of a flag that defaults to true, disabling the feature, something like a flag as "EnableSerialReloads = false" as default, and then if someone wants to enable it, set to true.

Booleans can become hard to define behavior (and this is why a bunch of APIs uses pointers, so you have true, false and undefined) so at least it is clear that if someone wants to Enable this feature, they use something like an "enable-something" flag.

This is better also for helm charts, etc.

Thanks!

rsafonseca commented:
Hi @rikatz, thanks for reviewing :)

Regarding concurrency and using a single variable, i don't think this would be a problem in the way things are designed overall, but let's see if anyone else has any concern here. I don't think it's worth the trouble to create a channel just for this, as the existing queue system already has a re-queue mechanism which is being employed here.

Regarding using a positive behaviour flag, that is a good point! I've refactored it and used your name suggestion there and added in a helm chart option for it as well :)

tao12345666333 commented:
/assign

Thanks, I will take a look this week.

tao12345666333 commented:
Sorry for the long delay. Let me take a look

strongjz commented on internal/ingress/controller/nginx.go:
```diff
@@ -87,9 +88,10 @@ func NewNGINXController(config *Configuration, mc metric.Collector) *NGINXContro
+		resolver:         h,
+		cfg:              config,
+		syncRateLimiter:  flowcontrol.NewTokenBucketRateLimiter(config.SyncRateLimit, 1),
+		workersReloading: false,
```
Shouldn't the default be set to config.WorkerSerialReloads

rsafonseca replied:
No. This is a state flag, not a config flag

rsafonseca replied:
The workersReloading flag controls the state of whether there is a reload going on, or not, and then if WorkerSerialReloads is enabed, and workersReloading is true, it will not start a new reload, but rather re-queue it. From the start, there is no reload happening so this is set to false.

rsafonseca commented:
> Sorry for the long delay. Let me take a look 

Any news @tao12345666333 ? I'm waiting to get this merged so I can move off my fork and upgrade ;)

tao12345666333 commented:
Thank you for your contributions!
I will finish the review this week.

BTW this PR need rebase

tao12345666333 commented:
/retest

tao12345666333 commented on internal/ingress/controller/nginx.go:
```diff
@@ -673,6 +677,11 @@ Error: %v
 //
 //nolint:gocritic // the cfg shouldn't be changed, and shouldn't be mutated by other processes while being rendered.
 func (n *NGINXController) OnUpdate(ingressCfg ingress.Configuration) error {
+	workerSerialReloads := n.store.GetBackendConfiguration().WorkerSerialReloads
+	if workerSerialReloads && n.workersReloading {
+		return errors.New("worker reload already in progress, requeuing reload")
+	}
+
 	cfg := n.store.GetBackendConfiguration()
```
We can just use the `cfg` instead of call it twice.

rsafonseca replied:
done :)

tao12345666333 commented on charts/ingress-nginx/values.yaml:
```diff
@@ -93,6 +93,8 @@ controller:
   # Global snippets in ConfigMap are still respected
   allowSnippetAnnotations: false
+  # -- This configuration defines if NGINX workers should reload serially instead of concurrently when multiple changes that require reloads are queued
+  enableSerialReloads: false
```
Let's change this to `enableWorkerSerialReloads` to make it more clearly. Thanks

rsafonseca replied:
done :)

rikatz commented:
@rsafonseca as we have bumped a bunch of libraries, please rebase over main to be sure codegen job will run using golang v1.22 and the latest k8s versions during tests :) 

tks

rsafonseca commented:
> @rsafonseca as we have bumped a bunch of libraries, please rebase over main to be sure codegen job will run using golang v1.22 and the latest k8s versions during tests :)
> 
> tks

I did rebase 5 days ago @rikatz and my branch is still up to date with main. could that job be pulling from the main in my fork instead? I handn't synced that one, since it wasn't the branch i was merging, but i synced it now

rsafonseca commented on internal/ingress/controller/nginx.go:
```diff
@@ -738,9 +747,41 @@ func (n *NGINXController) OnUpdate(ingressCfg ingress.Configuration) error {
+	klog.V(3).Infof("waiting for worker count to be equal to %s", expectedWorkers)
+	for numWorkers != expectedWorkers {
+		time.Sleep(time.Second)
+		o, err := exec.Command("/bin/sh", "-c", "pgrep worker | wc -l").Output()
```
It's been working well here for a few months now, in ~30+ clusters and no issues so far :)
If you find any issue do let me know :D

tao12345666333 commented:
/test pull-ingress-nginx-codegen

tao12345666333 commented:
I have created a PR https://github.com/kubernetes/ingress-nginx/pull/11344 
It will unblock the failures of the codegen job.

tao12345666333 commented:
#11344 has been merged, let's do a rebase and the codegen will be passed

rsafonseca commented:
Looks good @tao12345666333 , thanks for fixing this in main :)

rsafonseca commented:
Any blockers for merging this now? Can we get it shipped? :)

tao12345666333 reviewed (commented):
/lgtm

tao12345666333 commented:
Thank you for your contributions! @rsafonseca

bmv126 commented:
If there are multiple updates and now with this feature flag set to `true` once all the reload ends, will nginx be running with latest configuration ?
The query is because of this code which looks like blocks further reload.
https://github.com/kubernetes/ingress-nginx/blob/main/internal/ingress/controller/nginx.go#L685


Can you clarify here ?
